modelmapper/loader.py

```diff
@@ -9,38 +10,121 @@ def select(fields):
+    Base class for loaders.
+    """
+    def pre_row_insert(self, row, session, table):
+        """Override to add any logic for rows before being loaded"""
+        return row
+
+    def post_row_insert(self, row, session, table):
+        """Override to add any logic for rows after being loaded"""
+        pass
+
+    def insert_row_into_db(self, row, session, table):
+        """REQUIRED IMPLEMENTATION for how a row is loaded into the db"""
+        raise NotImplementedError("Please implement the insert_row_into_db.")
+
+    def insert_chunk_of_data_to_db(self, session, table, chunk):
+        """Processing chunks of rows to be loaded into db"""
+        count = 0
+        if chunk:
+            for row in chunk:
+                if row is None:
+                    continue
+                row = self.pre_row_insert(row, session, table)
+                if row is not None:
+                    self.insert_row_into_db(row, session, table)
+                    self.post_row_insert(row, session, table)
+                    count += 1
+        return count
+
+
+class SimpleSqlalchemyLoaderMixin(BaseLoaderMixin):
+    def insert_row_into_db(self, row, session, table):
+        """Insert row into db"""
+        ins = table.insert().values(**row)
+        try:
+            session.execute(ins)
+            session.flush()
+        except Exception as e:
+            self.logger.error(f"Error on inserting row of data: {row}")
+            raise e
+
+
+class BaseSignatureSqlalchemyMixin(BaseLoaderMixin):
+    all_recent_rows_signatures = set()
+
+    def add_row_signature(self, chunk):
+        for row in chunk:
+            row['signature'] = signature = generate_row_signature(
+                row, self.RECORDS_MODEL, self.settings.ignore_fields_in_signature_calculation
+            )
+            if signature and signature not in self.all_recent_rows_signatures:
+                self.all_recent_rows_signatures.add(signature)
+                yield row
+
+    def get_id_by_signature(self, session, table, signature):
+        """Searches given table for given signature. Returns row if found or None if not"""
+        query = select([table.c.id]).where(table.c.signature == signature)
+        query_result = session.execute(query)
+        result = query_result.fetchone()
+        result = None if result is None else result[0]
+        return result
+
+    def insert_chunk_of_data_to_db(self, session, table, chunk):
+        """Add row signature to row then run Base class logic"""
+        new_chunk = self.add_row_signature(chunk)
+        return super(BaseSignatureSqlalchemyMixin, self).insert_chunk_of_data_to_db(session, table, new_chunk)
+
+
+class UniqueSignatureSqlalchemyLoaderMixin(BaseSignatureSqlalchemyMixin):
+    """
+    Sqlalchemy Specific Loader that backs up raw data into S3.
+    It appends a signature to the row only uploads if it is not already in the table.
+    Requires add_row_signature() method to be defined in the class.
```

Comment:
This is already included by default now, yes?

Also, the line above:
`It appends a signature to the row _and_ only uploads if it is not already in the table.`

Reply:
:+1